### ISSUE TITLE
[Feature Request] - Populate ratings from file metadata

### ISSUE THREAD

**Issue by gr34n** [navidrome/navidrome#1301]
I have a considerable song collection, which I started to rate at some point to organize it. All web players I've tried so far supported only their own rating system in their own databases. I never really bothered to transfer ratings there because of enormous amount of manual work and lack of compatibility with another players (in case of migration).

As I see it, the main purpose of web players is to access and parse existing song collections on the go. It'd be consistent to integrate song ratings from metadata so they can be used for sorting songs and creating smartlists right after scanning.

The only issue I see here is possible reverting of ratings changed via web player during subsequent scans. To avoid it rating scanning could be triggered only as a separate command. Another workaround is to add an option to block any changes and read ratings strictly from files (which is more drastic imo). I don't think it's worth the hassle to write web-changed ratings back to metadata.

**Comment by certuna** [comment]
Yeah, you're touching on a number of ratings (and playcounts!) issues.

- if Navidrome stays read-only (which I hope so, it's a good thing from a security and file corruption point of view), we cannot write our playcounts and ratings to the tags
- reading ratings from tags (`POPM` in id3): ratings are [stored by-user](https://id3.org/id3v2.4.0-frames) (specified by email address!), how should navidrome read them if these email accounts don't match the ND usernames?
- reading ratings/playcounts from tags: only once (on import) or whenever they're changed by other applications?
- if we're reading from tags but not writing our own, then we need to keep two counters/ratings fields: one that contains the value in the tag, and one field that Navidrome writes internally.
- in that case, rating in Navidrome _replaces_ the tag value, with playcount the Navidrome play count is _added_ to tag value.
- if we at some point add the possibility to import ratings, likes/stars/hearts and play counts from /r/Plex, /r/AppleMusic (/iTunes), /r/Airsonic, /r/Emby, etc, we need to create a new column in the DB for each of these applications.

**Comment by gr34n** [comment]
As I said, I don't think ratings are reason enough to switch files to RW-mode.

Ok, I forgot about email part of id3 specs. Still, doesn't seem too difficult. It can be an optional field in personal settings or in dialog box in case of manual import.

While I can't say anything against adding second rating field to the DB, displaying/managing several ratings in web player itself can make UI more cumbersome. And how mobile clients will understand which rating to show? Keeping only one rating field sounds more simple and reliable. The only question is what options do we have to populate that field.

1. (current) Ratings can be set only manually and not synced at all (even one way) with existing metadata. Obvious drawback here is the amount of time required to map thousands of songs basically duplicating existing data. I assume no one will do that and hence fully use rating system.
2. A toggle to read ratings strictly from metadata and disable their online editing. While it offers regular syncs, it can be useful only for people who manage their collection locally. It also may require some workarounds for mobile clients.
3. An option to run one-time/on demand import from metadata. This approach doesn't offer synchronization, yet it solves blank rating fields issue and doesn't introduce restrictions. Afterwards we can freely edit ratings or leave them as is. This is the most universal option.

For now I don't see any other plausible variants.

Reading from other applications is a beast of a different kind (even more complex I'd say). Apparently libraries should be somehow matched minding all subtle possible differences in titles/albums/artists like missing commas, apostrophes, quotes, etc. Sounds like overkill.
Also raises the question, what is NV? All those integrations can be justified if NV strives to become something like 'personal media data collector/analyzer' (does it really?). For 'web player' focusing on streaming and parsing private collections is more important (and we stream our files, not AppleMusic tracks via NV, duh :relaxed: ).

I'd appreciate if NV allowed ratings to be imported and included in smartlists at some point. Sadly, I lack skills to directly contribute to the project. Well, NV uses SQLite unlike other players I've seen, so at least I'll try to write a script to import ratings directly as a workaround.

**Comment by certuna** [comment]
> While I can't say anything against adding second rating field to the DB, displaying/managing several ratings in web player itself can make UI more cumbersome. 

I might not have been clear enough - obviously we're not going to show multiple ratings in the UI, this storage of different ratings would only be serverside, the API just exposes one rating clientside.

> For now I don't see any other plausible variants.

Sure there is:
1. import ratings from metadata and write to a `tag_rating` DB field
2. allow rating from the navidrome web UI or Subsonic clients, written to `navidrome_rating`
3. the API has a merit order so Navidrome ratings take preference over tagged ratings
4. after you've rated some songs in another applications that writes ratings to tags, you can re-sync ratings, since you won't overwrite the `navidrome_rating` field - that's the whole reason to store it in two different DB fields

> Apparently libraries should be somehow matched minding all subtle possible differences in titles/albums/artists like missing commas, apostrophes, quotes, etc. Sounds like overkill.

You would simply match on filepath, that's relatively easy. That's how Plex does their import.

> and we stream our files, not AppleMusic tracks via NV, duh

It's a bit confusing these days, but Apple Music the current Apple application for local music (the successor to iTunes). The streaming service is _also_ called Apple Music, but obviously that has nothing to do with your own local music collection.

**Comment by deluan** [comment]
I think importing and maintaining ratings from tags would be out of scope for Navidrome, I don't even know of any players that writes ratings in tags. Having said that, this could be easily accomplished by writing a script to import the ratings. Such script could write to Navidrome's DB using one of the two methods:
1) Directly to the SQLite3 DB (table `annotation`)
2) Using Subsonic API: can be more complex to implement but would work for other compatible Subsonic servers

**Comment by certuna** [comment]
Yeah the whole concept of writing ratings (and playcounts) in tags was part of the original specification of id3 but was never really adopted in practice.

Downside is that we now have to migrate ratings and playcounts database-to-database.

**Comment by gr34n** [comment]
> The more I think of it, this actually might be a good reason to consider supporting reading ratings/playcounts from tags, by the way. For a migration, you'd only have to write scripts that write Plex/AppleMusic/Kodi/etc ratings-to-tags, and then Navidrome would simply pick up the tagged ratings/playcounts.

Yeah, metadata is the only at least remotely standardized ratings implementation I've come up with during past migrations. Without some common denominator you'll be bound by one player.

> I think importing and maintaining ratings from tags would be out of scope for Navidrome, I don't even know of any players that writes ratings in tags. 

Web and mobile players do tend to use some custom ratings system. Desktop players usually work with id3, and still there are few mature mobile players with id3 ratings support.

As for the script, it's already done (thanks for the tips, though) and ratings are inserted into database. Now I'm waiting for smartlist support to be able to do something with them 😆

**Comment by deluan** [comment]
> As for the script, it's already done (thanks for the tips, though) and ratings are inserted into database. Now I'm waiting for smartlist support to be able to do something with them 😆

Great! No worries, it is coming... :)

**Comment by deluan** [comment]
@certuna:

> The more I think of it, this actually might be a good reason to consider supporting reading ratings/playcounts from tags, by the way. For a migration, you'd only have to write scripts that write Plex/AppleMusic/Kodi/etc ratings-to-tags, and then Navidrome would simply pick up the tagged ratings/playcounts.

If you have to write a script to get the ratings into tags, you can write the script to write it directly to ND DB....

**Comment by quietmansoath** [comment]
So in advance for the long post. But I would like to propose a reconsideration of the decisions to keep Navidrome read only and not read user tags/ID3 tags for ratings (and other possible information).

Unless a user wants to use iTunes, there doesn't seem to be a single solution today that will allow them to listen to their own music on multiple devices, and manage the data in that collection.

Instead, there are just various schema for database lock-in. I've tried iTunes, Plex, all of the music players available on linux (yes), all of the music players available on android (yes, all), and some on Windows. MusicBee, MediaMonkey, Windows Media Player (on windows), and Gone Mad Media Player and PowerAmp (on Android), are the only ones that at least allow writing of rating metadata to the file, or a conversion from a database to this (I guess RocketPlayer for android also, technically, if you want to stay in XML architecture). 

There may be a couple more, but the point I'm trying to make is that, like Navidrome (currently), most music players do not. The result is a very fractured music player environment where information isn't really compatible with anything else out there and it makes for a extremely painful user experience if and when they decide to switch. Search for almost any music player and "migrate ratings" or "doesn't read ratings" and you'll see plenty of long forum threads "addressing" the problem that eventually end in "our app doesn't support that and won't."

So to repeat @gr34n 's question, what is Navidrome? 

- Is it just another beautiful web player to replace subsonic's sometimes buggy and feature lacking code?
- Or are there plans to eventually make it a "personal media data collector/analyzer"? (Which is part of what has made MusicBee on Windows so popular, and what led to iTunes' initial popularity, by the way)

I ask this and bring up these points because I'm interested in creating a music player for android that gets along with a subsonic api-reading web or desktop player that will better allow music portability by updating the file's metadata, but if those players don't read file metadata and refuse to update file metadata, then making an android app is pointless. And I'm not a developer by any means, but this problem has been frustrating enough over the last decade that I've already started trying to learn.

Alternatively, if there was API (subsonic or Navidrome-specific) or other functionality built into Navidrome from the get-go to allow for synced metadata management in some capacity, it would be extremely helpful in allowing others to continue to expand on this project. (Though I personally prefer ID3 tag/file updating as otherwise we're just recreating the iTunes lock-in problem).

Some other questions for @certuna :

- What are the security issues involved in making Navidrome read-write?
- Also as for the "**Reading from other applications' database**" problem:

> - so you'd have in our `mediafile` table the following fields: `tag_playcount`, `apple_playcount`, `plex_playcount`, `airsonic_playcount`, `emby_playcount` and `navidrome_playcount`. The Playcount displayed in the Navidrome UI will be the sum of all these.
> - same with Ratings: you'd store the ratings from each of these applications, plus the Navidrome ratings. What we show in the UI is then probably a cascade, i.e. Navidrome rating trumps Plex rating trumps tag rating, or something.
> - then, after you've played a song in Plex, you (manually?) resync the playcounts in Navidrome, and `plex_playcount` gets updated.
> - an import-from-other-apps function is not a trivial thing - you'd need to point the scanner to each of those apps's SQLite database (or in the case of AppleMusic, the XML), then cycle through each file, pull out the filepath/rating/heart/playcount, look up each filepath in the Navidrome db, and write the appropriate values. It will be slow.

—Excuse my ignorance here, but I don't see why all of these databases would need to be "active" at the same time. Couldn't something be setup in the user settings so that they could "turn off" the applications they don't care about? So if a user is migrating from plex to navidrome for example, they could uncheck boxes that would contol whether the app is maintaining writing the files in a way Emby needs to read them. Not sure if that makes sense, but wouldn't that potentially speed things up?

**Comment by quietmansoath** [comment]
One last small thing to point out (sorry):

Navidrome currently advertises that it "Reads and uses all your beautifully curated metadata" but if that doesn't count ratings and playcounts that [a user could painstakingly add in metadata management software like MP3 tag](https://help.mp3tag.de/main_tags.html), the advertised statement is...not fully accurate.
